File: src/main/problems/java/Nr23.java
package main.problems.java;

import main.util.java.ListNode;

public class Nr23 {
	public ListNode mergeKLists(ListNode[] lists) {
		if (lists == null || lists.length == 0) {
			return null;
		}

		for (int i = 1; i < lists.length; i++) {
			lists[0] = mergeTwoLists(lists[0], lists[i]);
		}

		return lists[0];
	}

	public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
		if (l1 == null) return l2;
		if (l2 == null) return l1;

		if (l1.val < l2.val) {
			l1.next = mergeTwoLists(l1.next, l2);
			return l1;
		}
		l2.next = mergeTwoLists(l1, l2.next);
		return l2;
	}
}

File: src/main/problems/ruby/Nr23.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode[]} lists
# @return {ListNode}
def merge_k_lists(lists)
    return lists if lists.size == 0
    return lists[1] if lists.size == 1
    
    merge_lists(lists, 0, lists.size - 1)
end

def merge_lists(lists, first, last)
    return lists[first] if first == last
    return merge_two_lists(lists[first], lists[last]) if last - first == 1

    pivot = (first +  last) / 2
    return merge_two_lists(
        merge_lists(lists, first, pivot),
        merge_lists(lists, pivot + 1, last)
    )
end

def merge_two_lists(list1, list2)
    return list2 unless list1
    return list1 unless list2

    if list1.val < list2.val
        list1.next = merge_two_lists(list1.next, list2)
        list1
    else
        list2.next = merge_two_lists(list1, list2.next)
        list2
    end
end